NE-969: Add IBM DNS Services provider support for private clusters

SzucsAti opened the pull request:
With this PR, cluster-ingress-operators running on Private IBMCloud IPI/UPI clusters will be able to use DNS Services (https://cloud.ibm.com/docs/dns-svcs?topic=dns-svcs-about-dns-services) as their DNS provider.

Pulling in latest openshift/api changes as it's needed for this PR.

This PR is required for Private IBMCloud IPI/UPI clusters.

jeffnowicki commented:
/cc @Miciah

gcs278 commented:
/assign @gcs278

gcs278 commented:
Thanks for the PR. We discussed this in our PR review meeting. I'll do a review sometime next week.

gcs278 commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+	return kerrors.NewAggregate(errs)
+}
+
+func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1.DNSZone) error {
```
*nit* does zone need to be pass-by-value here? Doesn't that do an unnecessary memory copy?
Same comment goes for `cis_provider.go` to be consistent.

SzucsAti replied:
Changed it. (Side note: as I see this change is inconsistent with the way other platform providers do it.)

gcs278 replied:
@SzucsAti yea reviewing this again, I was asking the question, not necessarily suggesting it to be changed. Sorry I should have been more clear with my statement that I wasn't 100% sure.

I was originally thinking pointers are just overall better, but after reading https://medium.com/@meeusdylan/when-to-use-pointers-in-go-44c15fe04eac, it really only matters when there are large complex structures.

If you see this change as inconsistent, I'd recommend changing back. Sorry about that.

SzucsAti replied:
Reverted the change.

gcs278 commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+	return nil
+}
+
+func validateInputDNSData(record *iov1.DNSRecord, zone configv1.DNSZone) error {
```
*nit* does zone need to be pass-by-value here? Doesn't that do an unnecessary memory copy?
Same comment goes for `cis_provider.go` to be consistent.

SzucsAti replied:
Done.

gcs278 reviewed (commented):
Looks pretty good, thanks for the PR. I'm have a small concern with code reuse and consistency here. Seems code was adapted from existing code and there are a few duplicate logic sections. I'm just wonder if we could structure to enable better code reuse and less unique logic paths. I will get back to you on maybe a potential suggestion, but at the moment, I don't have any other suggestions other than my comments.

SzucsAti commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -650,14 +661,30 @@ func getIbmDNSProvider(dnsConfig *configv1.DNS, creds *corev1.Secret, cisInstanc
+		if err != nil {
+			return nil, fmt.Errorf("failed to create IBM CIS DNS manager: %v", err)
+		}
+		log.Info("successfully initialized ibm cis dns provider")
```
Fixed.

SzucsAti commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -650,14 +661,30 @@ func getIbmDNSProvider(dnsConfig *configv1.DNS, creds *corev1.Secret, cisInstanc
+		if err != nil {
+			return nil, fmt.Errorf("failed to create IBM DNSSVCS DNS manager: %v", err)
+		}
+		log.Info("successfully initialized ibm dnssvcs dns provider")
```
Fixed.

SzucsAti commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -650,14 +661,30 @@ func getIbmDNSProvider(dnsConfig *configv1.DNS, creds *corev1.Secret, cisInstanc
+		log.Info("successfully initialized ibm cis dns provider")
+		return provider, nil
+	} else {
+		provider, err := ibmprivatedns.NewProvider(ibmprivatedns.Config{
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider_test.go:
```diff
@@ -0,0 +1,285 @@
+		recordedCall                 string
+		listAllDnsRecordsInputOutput dnsclient.ListAllDnsRecordsInputOutput
+		updateDnsRecordInputOutput   dnsclient.UpdateDnsRecordInputOutput
+		expectedErr                  bool
```
Makes sense. Added expectErrorContains and updated the tests accordingly.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+		}
+	}
+	if result == nil {
+		return fmt.Errorf("delete: invalid result")
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+			}
+		}
+		if listResult == nil {
+			return fmt.Errorf("createOrUpdateDNSRecord: invalid result")
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+
+	options := &dnssvcsv1.DnsSvcsV1Options{
+		Authenticator: authenticator,
+		URL:           "https://api.dns-svcs.cloud.ibm.com/v1",
```
Moved them to constants.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+	if err := validateDNSServices(provider); err != nil {
+		return nil, fmt.Errorf("failed to validate ibm dns services: %w", err)
+	}
+	log.Info("Successfully validated DNSServices")
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+
+	// TTL must be one of [1 60 120 300 600 900 1800 3600 7200 18000 43200]
+	useDefaultTTL := true
+	for _, v := range []int64{1, 60, 120, 300, 600, 900, 1800, 3600, 7200, 18000, 43200} {
```
These TTL constraints come from the actual dns services. Unfortunately they are inconsistent.

In case of CIS, TTL must be between 120 and 2,147,483,647 seconds, or 1 for Automatic. The code makes sure that in case record.Spec.RecordTTL has an invalid value (0 or 2-119) then the default TTL value will be used.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+	return nil
+}
+
+func validateInputDNSData(record *iov1.DNSRecord, zone configv1.DNSZone) error {
```
Couldn't really find a place to put these common functions/structures, so ended up creating common.go in the dns/ibm folder. Moved validateInputDNSData there to be used by both providers.

gcs278 replied:
That sounds like a step in the right direction. @Miciah should be doing a second-pass review, and I'm curious to see what he thinks about structure.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+	"time"
+
+	"github.com/IBM/go-sdk-core/v5/core"
+	dnssvcsv1 "github.com/IBM/networking-go-sdk/dnssvcsv1"
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+)
+
+type Provider struct {
+	dnsService dnsclient.DnsClient
```
In case of CIS sdk, zoneID is stored and used in the provider options which is set during client creation. So we decided to create a zoneID to client map, so we don't have to create a client each time there is a dns operation. https://github.com/SzucsAti/cluster-ingress-operator/blob/ibm-dnssvc-support/pkg/dns/ibm/public/cis_provider.go#L45-#L52

In case of DNS-SVCS sdk, zoneID is a required parameter in every function and it's not needed to set in client options. https://github.com/SzucsAti/cluster-ingress-operator/blob/ibm-dnssvc-support/pkg/dns/ibm/private/dnssvcs_provider.go#L44-#L49

gcs278 replied:
Got it, thanks for the explanation.

SzucsAti commented:
@gcs278 Thank you for the review. Addressed your comments. 

I agree that the usage of the providers looks similar. Unfortunately there are noticeable differences when it comes to the providers' sdks, which makes it hard to reuse code for both providers without overengineering, but I'm open to suggestions. I created a common.go file and moved duplicated config/function there based on your comments.

gcs278 reviewed (commented):
thanks for the comment responses. Looking good.

SzucsAti commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -640,8 +652,8 @@ func (r *reconciler) customCABundle() (string, error) {
 
-//getIbmDNSProvider will intialize the IBM DNS provider
-func getIbmDNSProvider(dnsConfig *configv1.DNS, creds *corev1.Secret, cisInstanceCRN, userAgent string) (*ibmdns.Provider,
+//getIbmDNSProvider will intialize an IBM DNS provider
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/client/fake_client.go:
```diff
@@ -0,0 +1,124 @@
+	"errors"
+
+	"github.com/IBM/go-sdk-core/v5/core"
+	dnssvcsv1 "github.com/IBM/networking-go-sdk/dnssvcsv1"
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/client/client.go:
```diff
@@ -0,0 +1,23 @@
+
+import (
+	"github.com/IBM/go-sdk-core/v5/core"
+	dnssvcsv1 "github.com/IBM/networking-go-sdk/dnssvcsv1"
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,246 @@
+			}
+		}
+		if listResult == nil {
+			return fmt.Errorf("validateDNSServices: NewListResourceRecordsOptions returned nil as result")
```
Right, this check is not needed. Realised that errors were not aggregated, just returned.. fixed that as well.

gcs278 commented:
@Miciah I've done my most of my reviewing, mind taking a look when you get a chance?

gcs278 commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,273 @@
+	config     Config
+}
+
+type Config struct {
```
I think that makes sense to me. I see `InstanceID` is a bit overloaded since it could be different types of ID's for public or private (CRN ID vs. IstanceID?). But I think this is okay.

gcs278 commented:
I took another pass. The I did miss your comment response about the pointers, but I agree with your comment about inconsistency. I think we should revert those pointer updates.

Otherwise just waiting on a second pass from Miciah.

Miciah commented on pkg/dns/ibm/common.go:
```diff
@@ -0,0 +1,40 @@
+	iov1 "github.com/openshift/api/operatoringress/v1"
+)
+
+// Common configuration of ibm providers
+// InstanceID is GUID in case of dns svcs and CRN in case of cis
+type Config struct {
+	APIKey     string
+	InstanceID string
+	UserAgent  string
+	Zones      []string
```
This would be more Go idiomatic:
```suggestion
// Config holds common configuration of IBM DNS providers.
type Config struct {
	// APIKey is the IBM Cloud API key that the provider will use.
	APIKey     string
	// InstanceID is GUID in case of dns svcs and CRN in case of cis.
	InstanceID string
	// UserAgent is the user-agent identifier that the client will use in
	// HTTP requests to the IBM Cloud API.
	UserAgent  string
	// Zones is a list of DNS zones in which the DNS provider manages DNS records.
	Zones      []string
```

SzucsAti replied:
Thank you. Applied your suggestion.

Miciah commented on pkg/dns/ibm/common.go:
```diff
@@ -0,0 +1,40 @@
+	Zones      []string
+}
+
+func ValidateInputDNSData(record *iov1.DNSRecord, zone *configv1.DNSZone) error {
```
Please add a godoc comment.

SzucsAti replied:
Done.

Miciah commented on pkg/dns/ibm/private/client/client.go:
```diff
@@ -0,0 +1,23 @@
+	"github.com/IBM/networking-go-sdk/dnssvcsv1"
+)
+
+type DnsClient interface {
+	NewListResourceRecordsOptions(instanceID string, dnszoneID string) *dnssvcsv1.ListResourceRecordsOptions
+	ListResourceRecords(listResourceRecordsOptions *dnssvcsv1.ListResourceRecordsOptions) (result *dnssvcsv1.ListResourceRecords, response *core.DetailedResponse, err error)
+	NewDeleteResourceRecordOptions(instanceID string, dnszoneID string, recordID string) *dnssvcsv1.DeleteResourceRecordOptions
+	DeleteResourceRecord(deleteResourceRecordOptions *dnssvcsv1.DeleteResourceRecordOptions) (response *core.DetailedResponse, err error)
+	NewUpdateResourceRecordOptions(instanceID string, dnszoneID string, recordID string) *dnssvcsv1.UpdateResourceRecordOptions
+	NewResourceRecordUpdateInputRdataRdataCnameRecord(cname string) (_model *dnssvcsv1.ResourceRecordUpdateInputRdataRdataCnameRecord, err error)
+	NewResourceRecordUpdateInputRdataRdataARecord(ip string) (_model *dnssvcsv1.ResourceRecordUpdateInputRdataRdataARecord, err error)
+	UpdateResourceRecord(updateResourceRecordOptions *dnssvcsv1.UpdateResourceRecordOptions) (result *dnssvcsv1.ResourceRecord, response *core.DetailedResponse, err error)
+	NewCreateResourceRecordOptions(instanceID string, dnszoneID string) *dnssvcsv1.CreateResourceRecordOptions
+	NewResourceRecordInputRdataRdataCnameRecord(cname string) (_model *dnssvcsv1.ResourceRecordInputRdataRdataCnameRecord, err error)
+	NewResourceRecordInputRdataRdataARecord(ip string) (_model *dnssvcsv1.ResourceRecordInputRdataRdataARecord, err error)
+	CreateResourceRecord(createResourceRecordOptions *dnssvcsv1.CreateResourceRecordOptions) (result *dnssvcsv1.ResourceRecord, response *core.DetailedResponse, err error)
+	NewGetDnszoneOptions(instanceID string, dnszoneID string) *dnssvcsv1.GetDnszoneOptions
+	GetDnszone(getDnszoneOptions *dnssvcsv1.GetDnszoneOptions) (result *dnssvcsv1.Dnszone, response *core.DetailedResponse, err error)
+}
```
Seems like this interface should be defined in `"github.com/IBM/networking-go-sdk/dnssvcsv1"`, but I guess we can't do anything about that.  Maybe add a comment to say that we're defining this interface to describe the methods defined in `dnssvcsv1` for the purpose of having an interface that we can use in the implementation and test code for cluster-ingress-operator's provider logic.

SzucsAti replied:
Exactly. Added a comment.

Miciah commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+var (
+	_                       dns.Provider = &Provider{}
+	log                                  = logf.Logger.WithName("dns")
+	defaultDNSSVCSRecordTTL              = int64(120)
+	DNSSVCSApiEndpoint                   = "https://api.dns-svcs.cloud.ibm.com/v1"
+)
```
`defaultDNSSVCSRecordTTL` and `DNSSVCSApiEndpoint` can be declared as consts.  Since `DNSSVCSApiEndpoint` isn't used outside this package, it doesn't need to be exported.  Please add godoc!
```suggestion
)

const (
	// defaultDNSSVCSRecordTTL is the default TTL used when a DNS record
	// does not specify a valid TTL.
	defaultDNSSVCSRecordTTL = int64(120)
	// dnsSvcsApiEndpoint is the API endpoint for IBM Cloud DNS Services. 
	dnsSvcsApiEndpoint      = "https://api.dns-svcs.cloud.ibm.com/v1"
)
```
Actually, could you use `dnssvcsv1.DefaultServiceURL` and do away with `DNSSVCSApiEndpoint` entirely?

SzucsAti replied:
Using DefaultServiceURL.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+		if response != nil && response.StatusCode != http.StatusNotFound {
+			return fmt.Errorf("delete: failed to list the dns record: %w", err)
+		}
+	}
```
Agree, changed it.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+			}
+			if *resourceRecord.Type == string(iov1.ARecordType) {
+				resourceRecordTarget = fmt.Sprint(rData["ip"])
+			}
```
> Can we be absolutely certain that resourceRecord.Type is non-nil? I'd prefer this logic include a nil check.

There is a `ValidateInputDNSData(record, zone)` call in the beginning of the functions that does nil checks.

> What if the record has some other type? I'd prefer this logic use a switch with a default clause.

Makes sense. Changed it to switch-case.

> What are the possible types of rData["cname"] and rData["ip"]? Using fmt.Sprint here seems slightly odd.

It supposed to be string only. Added a type check just to be safe and changed the way we convert it.

Miciah replied:
> > Can we be absolutely certain that resourceRecord.Type is non-nil? I'd prefer this logic include a nil check.
> 
> There is a `ValidateInputDNSData(record, zone)` call at the beginning of the functions that does nil checks.

That checks the input that is used for the `ListResourceRecords` call, but it doesn't validate the result, right?  Is it safe to assume that the result is valid if the input was valid?

SzucsAti replied:
Oh correct. Added nil check validation for list result.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+				resourceRecordTarget = fmt.Sprint(rData["ip"])
+			}
+
+			if resourceRecordTarget == target && *resourceRecord.Name == DNSName {
```
There is a `ValidateInputDNSData(record, zone)` call at the beginning of the functions that does nil checks.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+				delOpt := p.dnsService.NewDeleteResourceRecordOptions(p.config.InstanceID, zone.ID, *resourceRecord.ID)
+				delResponse, err := p.dnsService.DeleteResourceRecord(delOpt)
+				if err != nil {
+					if delResponse != nil && delResponse.StatusCode != http.StatusNotFound {
```
I believe we should return error in that case. Changed it accordingly.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+		return fmt.Errorf("delete: ListResourceRecords returned nil as result")
+	}
+	for _, target := range record.Spec.Targets {
+		for _, resourceRecord := range result.ResourceRecords {
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+
+	listOpt := p.dnsService.NewListResourceRecordsOptions(p.config.InstanceID, zone.ID)
+	// Some dns records (e.g. wildcard record) have an ending "." character in the DNSName
+	DNSName := strings.TrimSuffix(record.Spec.DNSName, ".")
```
Renamed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+	}
+
+	listOpt := p.dnsService.NewListResourceRecordsOptions(p.config.InstanceID, zone.ID)
+	// Some dns records (e.g. wildcard record) have an ending "." character in the DNSName
```
It was referring to the way the ingress operator creates its wildcard record and users that might create records similarly. Applied your suggestion.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+
+	// TTL must be one of [1 60 120 300 600 900 1800 3600 7200 18000 43200]
+	useDefaultTTL := true
+	for _, v := range []int64{1, 60, 120, 300, 600, 900, 1800, 3600, 7200, 18000, 43200} {
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+		}
+		if listResult == nil {
+			return fmt.Errorf("createOrUpdateDNSRecord: ListResourceRecords returned nil as result")
+		}
```
Not necessary, moved it to out of the loop.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+				updateOpt := p.dnsService.NewUpdateResourceRecordOptions(p.config.InstanceID, zone.ID, *resourceRecord.ID)
+				updateOpt.SetName(DNSName)
+				if record.Spec.RecordType == iov1.CNAMERecordType {
+					inputRData, error := p.dnsService.NewResourceRecordUpdateInputRdataRdataCnameRecord(target)
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+
+	listOpt := p.dnsService.NewListResourceRecordsOptions(p.config.InstanceID, zone.ID)
+	// Some dns records (e.g. wildcard record) have an ending "." character in the DNSName
+	DNSName := strings.TrimSuffix(record.Spec.DNSName, ".")
```
Renamed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+	return kerrors.NewAggregate(errs)
+}
+
+func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone *configv1.DNSZone) error {
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+		for _, resourceRecord := range listResult.ResourceRecords {
+
+			if *resourceRecord.Name == DNSName {
+				updated = true
```
Agree. Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+					}
+					updateOpt.SetRdata(inputRData)
+				} else {
+					inputRData, error := p.dnsService.NewResourceRecordUpdateInputRdataRdataARecord(target)
```
Renamed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,241 @@
+			createOpt.SetType(string(record.Spec.RecordType))
+
+			if record.Spec.RecordType == iov1.CNAMERecordType {
+				inputRData, error := p.dnsService.NewResourceRecordInputRdataRdataCnameRecord(target)
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider_test.go:
```diff
@@ -0,0 +1,273 @@
+
+	dnsService, err := dnsclient.NewFake()
+	if err != nil {
+		t.Fatal("failed to create fakeClient")
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider_test.go:
```diff
@@ -0,0 +1,273 @@
+
+	dnsService, err := dnsclient.NewFake()
+	if err != nil {
+		t.Fatal("failed to create fakeClient")
```
Done.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider_test.go:
```diff
@@ -0,0 +1,273 @@
+
+			dnsService.DeleteDnsRecordInputOutput = tc.deleteDnsRecordInputOutput
+
+			err = provider.Delete(&record, zone)
+
+			if len(tc.expectErrorContains) != 0 && !strings.Contains(err.Error(), tc.expectErrorContains) {
+				t.Errorf("expected message to include %q, got %q", tc.expectErrorContains, err.Error())
+			}
```
Done.

SzucsAti commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -21,21 +22,15 @@ var (
 	_                   dns.Provider = &Provider{}
 	log                              = logf.Logger.WithName("dns")
 	defaultCISRecordTTL              = int64(120)
+	CISApiEndpoint                   = "https://api.cis.cloud.ibm.com/"
```
Using DefaultServiceURL instead.

SzucsAti commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -573,14 +576,23 @@ func (r *reconciler) createDNSProvider(dnsConfig *configv1.DNS, platformStatus *
+				return nil, fmt.Errorf("failed to create ibm cis dns provider: %v", err)
+			}
+		} else if platformStatus.IBMCloud.DNSInstanceCRN != "" {
+			rExp := regexp.MustCompile(`^crn:v1:(?P<environment>[a-z]+):public:(?P<type>[^:]+):(?P<region>[^:]+):a/(?P<account_id>[a-f0-9]+):(?P<guid>[a-f0-9\\-]+):(?P<resourceType>[a-z]*):(?P<resourceID>[a-f0-9\\-]*)$`)
```
Moved the definition to the ibm package. I was using a regexp from our internal repos, but looking at the official it doesn't have to be that strict. Example CRN of an actual service: `crn:v1:bluemix:public:dns-svcs:global:a/c9a69aeea7304d61b53dac4c76ff0e55:7c786ae5-3eeb-4a99-bd9c-35e199eb85b3::`

I think it could be added as a validator. Will discuss it with the colleagues.

SzucsAti commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -640,24 +652,37 @@ func (r *reconciler) customCABundle() (string, error) {
-//getIbmDNSProvider will intialize the IBM DNS provider
-func getIbmDNSProvider(dnsConfig *configv1.DNS, creds *corev1.Secret, cisInstanceCRN, userAgent string) (*ibmdns.Provider,
-	error) {
+//getIbmDNSProvider will initialize an IBM DNS provider
```
Fixed.

cjschaef commented:
@Miciah @SzucsAti Any further progress on getting this reviewed/approved?
This will soon become a blocker for the IBM Cloud IPI GA, once we start adding Private cluster support in the `installer` and I'm hoping to avoid that if possible.

SzucsAti commented:
Created an IPI cluster on IBMCloud and verified that the changes we made don't break the existing CIS integration. DNS Services integration seems to work properly as well. Waiting if there are more suggestions or concerns.

jeffnowicki commented:
/retest-required

SzucsAti commented:
Resolved the conflicts. We no longer need to update the `openshift/api` dependency as it was updated on the master branch.

jeffnowicki commented:
/retest-required

jeffnowicki commented:
I'm unsure why `e2e-aws-operator` and `e2e-gcp-serial` are set as required for this `ibmcloud` specific PR. May need to provide skips/overrides for those tests so CI bot can merge this PR (after required labels have been added).

gcs278 commented:
I took another look and it appears all comments have been addressed. @Miciah should lgtm as well and approve when ready.
/lgtm

Miciah commented:
> I'm unsure why `e2e-aws-operator` and `e2e-gcp-serial` are set as required for this `ibmcloud` specific PR. May need to provide skips/overrides for those tests so CI bot can merge this PR (after required labels have been added).

These tests run on every PR.  I see that e2e-aws-operator failed with the following:
```
=== RUN   TestAll/serial/TestHstsPolicyWorks
    hsts_policy_test.go:72: created a RequiredHSTSPolicy with DomainPatterns: [*.apps.ci-op-bh293p3p-43abb.origin-ci-int-aws.dev.rhcloud.com],
         preload policy: RequirePreload,
         includeSubDomains policy: RequireIncludeSubDomains,
         largest age: 99999,
         smallest age: 0
    hsts_policy_test.go:93: failed to create pod &Namespace{ObjectMeta:{hsts-policy-namespace    38ba4198-194a-4723-9992-553a88fd0347 42209 0 2022-08-15 14:28:59 +0000 UTC <nil> <nil> map[kubernetes.io/metadata.name:hsts-policy-namespace] map[] [] []  [{e2e.test Update v1 2022-08-15 14:28:59 +0000 UTC FieldsV1 {"f:metadata":{"f:labels":{".":{},"f:kubernetes.io/metadata.name":{}}}} }]},Spec:NamespaceSpec{Finalizers:[kubernetes],},Status:NamespaceStatus{Phase:Active,Conditions:[]NamespaceCondition{},},}/hsts-policy-echo: pods "hsts-policy-echo" is forbidden: violates PodSecurity "restricted:v1.24": allowPrivilegeEscalation != false (container "echo" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "echo" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "echo" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "echo" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
```
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/796/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1559166191245725696

We'll need a fix like https://github.com/openshift/cluster-dns-operator/pull/336 for cluster-ingress-operator.

Miciah commented on pkg/dns/ibm/common.go:
```diff
@@ -0,0 +1,51 @@
+	iov1 "github.com/openshift/api/operatoringress/v1"
+)
+
+var (
+	// IBMResourceCRNRegexp is a regular expression of IBM resource CRNs
+	// See: https://cloud.ibm.com/docs/account?topic=account-crn
+	IBMResourceCRNRegexp = `^crn:v[0-9]:(?P<cloudName>[^:]*):(?P<cloudType>[^:]*):(?P<serviceName>[^:]*):(?P<location>[^:]*):(?P<scope>[^:]*):(?P<guid>[^:]*):(?P<resourceType>[^:]*):(?P<resourceID>[^:]*)$`
```
This can be a const:
```suggestion
const (
	// IBMResourceCRNRegexp is a regular expression of IBM resource CRNs
	// See: https://cloud.ibm.com/docs/account?topic=account-crn
	IBMResourceCRNRegexp = `^crn:v[0-9]:(?P<cloudName>[^:]*):(?P<cloudType>[^:]*):(?P<serviceName>[^:]*):(?P<location>[^:]*):(?P<scope>[^:]*):(?P<guid>[^:]*):(?P<resourceType>[^:]*):(?P<resourceID>[^:]*)$`
```
Alternatively, you can define a compiled regexp (in which case it must be a var):
```suggestion
var (
	// IBMResourceCRNRegexp is a regular expression of IBM resource CRNs
	// See: https://cloud.ibm.com/docs/account?topic=account-crn
	IBMResourceCRNRegexp = regexp.MustCompile(`^crn:v[0-9]:(?P<cloudName>[^:]*):(?P<cloudType>[^:]*):(?P<serviceName>[^:]*):(?P<location>[^:]*):(?P<scope>[^:]*):(?P<guid>[^:]*):(?P<resourceType>[^:]*):(?P<resourceID>[^:]*)$`)
```

SzucsAti replied:
Defined it as a compiled regexp var.

Miciah commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -573,14 +576,23 @@ func (r *reconciler) createDNSProvider(dnsConfig *configv1.DNS, platformStatus *
+				return nil, fmt.Errorf("failed to create ibm cis dns provider: %v", err)
+			}
+		} else if platformStatus.IBMCloud.DNSInstanceCRN != "" {
+			rExp := regexp.MustCompile(ibm.IBMResourceCRNRegexp)
```
Better to define a package variable for the compiled regexp to avoid recompiling it every time `createDNSProvider` is called.  You can define it at the top of the file, where `log` is currently defined:
```go
var (
	log = logf.Logger.WithName(controllerName)

	// ibmResourceCRNCompiledRegexp is a regular expression of IBM resource CRNs.
	ibmResourceCRNCompiledRegexp = regexp.MustCompile(ibm.IBMResourceCRNRegexp)
)
```
...and then...
```suggestion
			rExp := regexp.MustCompile(ibmResourceCRNCompiledRegexp)
```
Alternatively, define `ibm.IBMResourceCRNRegexp` as a compiled regexp (see my previous comment).

SzucsAti replied:
Defined it as a compiled regexp var.

Miciah commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+
+	// defaultDNSSVCSRecordTTL is the default TTL used when a DNS record
+	// does not specify a valid TTL.
+	defaultDNSSVCSRecordTTL = int64(120)
```
`defaultDNSSVCSRecordTTL` can be declared as a const.

SzucsAti replied:
Changed it to const.

Miciah commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+	provider.config.Zones = config.Zones
+
+	if err := validateDNSServices(provider); err != nil {
+		return nil, fmt.Errorf("failed to validateIBM Cloud DNS Services: %w", err)
```
```suggestion
		return nil, fmt.Errorf("failed to validate IBM Cloud DNS Services: %w", err)
```

SzucsAti replied:
Fixed.

Miciah commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -640,24 +652,37 @@ func (r *reconciler) customCABundle() (string, error) {
-//getIbmDNSProvider will intialize the IBM DNS provider
-func getIbmDNSProvider(dnsConfig *configv1.DNS, creds *corev1.Secret, cisInstanceCRN, userAgent string) (*ibmdns.Provider,
-	error) {
+//getIbmDNSProvider initializes and returns an IBM DNS provider instance.
```
Sorry for being so persnickety, but...
```suggestion
// getIbmDNSProvider initializes and returns an IBM DNS provider instance.
```

SzucsAti replied:
Fixed.

Miciah commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -573,14 +576,23 @@ func (r *reconciler) createDNSProvider(dnsConfig *configv1.DNS, platformStatus *
+		if platformStatus.IBMCloud.CISInstanceCRN != "" {
+			dnsProvider, err = getIbmDNSProvider(dnsConfig, creds, platformStatus.IBMCloud.CISInstanceCRN, userAgent, true)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create ibm cis dns provider: %v", err)
```
Looks like you're changing "IBM CIS DNS" to upper-case elsewhere, right?  Anyway, might as well change to using `%w` while you're updating this code.  
```suggestion
				return nil, fmt.Errorf("failed to create IBM CIS DNS provider: %w", err)
```
Actually, `getIbmDNSProvider` already wraps whatever error it returns with "failed to initialize IBM CIS DNS provider", so is it necessary to wrap the error here?  Might as well just return it as-is, no?
```suggestion
				return nil, err
```

SzucsAti replied:
Correct, changed it.

Miciah commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -573,14 +576,23 @@ func (r *reconciler) createDNSProvider(dnsConfig *configv1.DNS, platformStatus *
+			}
+			dnsProvider, err = getIbmDNSProvider(dnsConfig, creds, matches[rExp.SubexpIndex("guid")], userAgent, false)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create ibm dnssvcs dns provider: %v", err)
```
```suggestion
				return nil, fmt.Errorf("failed to create IBM Cloud DNS Services provider: %w", err)
```
or
```suggestion
				return nil, err
```
per my other comment.

SzucsAti replied:
Changed it.

Miciah commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -640,24 +652,37 @@ func (r *reconciler) customCABundle() (string, error) {
+	if isPublic {
+		provider, err := ibmpublicdns.NewProvider(providerCfg)
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize IBM CIS DNS provider: %v", err)
```
```suggestion
			return nil, fmt.Errorf("failed to initialize IBM CIS DNS provider: %w", err)
```

SzucsAti replied:
Fixed.

Miciah commented on pkg/operator/controller/dns/controller.go:
```diff
@@ -640,24 +652,37 @@ func (r *reconciler) customCABundle() (string, error) {
+	} else {
+		provider, err := ibmprivatedns.NewProvider(providerCfg)
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize IBM Cloud DNS Services provider: %v", err)
```
```suggestion
			return nil, fmt.Errorf("failed to initialize IBM Cloud DNS Services provider: %w", err)
```

SzucsAti replied:
Fixed.

Miciah commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+				updateOpt := p.dnsService.NewUpdateResourceRecordOptions(p.config.InstanceID, zone.ID, *resourceRecord.ID)
+				updateOpt.SetName(dnsName)
+
+				switch *resourceRecord.Type {
+				case string(iov1.CNAMERecordType):
```
By the way, if you agree that this code _technically should_ handle the case where we have an A record and want a CNAME record or vice versa but you believe it _isn't practically_ necessary and isn't worth the time to implement the logic to do so right now, I would be satisfied with a "TODO" comment.

SzucsAti replied:
I agree with your comment, but I don't think it's worth to implement it right now. Added a TODO comment.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+			if value, ok := rData["cname"].(string); ok {
+				resourceRecordTarget = value
+			} else {
+				return fmt.Errorf("delete: resource data has record with unknown rData cname type: %v ", err)
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+			if value, ok := rData["ip"].(string); ok {
+				resourceRecordTarget = value
+			} else {
+				return fmt.Errorf("delete: resource data has record with unknown rData ip type: %v ", err)
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+				return fmt.Errorf("delete: resource data has record with unknown rData ip type: %v ", err)
+			}
+		default:
+			return fmt.Errorf("delete: resource data has record with unknown type: %v", resourceRecord)
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+		}
+
+		for _, target := range record.Spec.Targets {
+			if resourceRecordTarget == target && *resourceRecord.Name == dnsName {
```
Makes sense. Applied the suggestion.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+				delOpt := p.dnsService.NewDeleteResourceRecordOptions(p.config.InstanceID, zone.ID, *resourceRecord.ID)
+				delResponse, err := p.dnsService.DeleteResourceRecord(delOpt)
+				if err != nil {
+					if (delResponse != nil && delResponse.StatusCode != http.StatusNotFound) || delResponse == nil {
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+
+	listResult, response, err := p.dnsService.ListResourceRecords(listOpt)
+	if err != nil {
+		if response != nil && response.StatusCode != http.StatusNotFound {
```
Fixed.

SzucsAti commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -0,0 +1,259 @@
+					}
+					updateOpt.SetRdata(inputRData)
+				default:
+					return fmt.Errorf("createOrUpdateDNSRecord: resource data has record with unknown type: %v", resourceRecord)
```
Fixed.

SzucsAti commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -1,43 +1,40 @@
+
+	// defaultCISRecordTTL is the default TTL used when a DNS record
+	// does not specify a valid TTL.
+	defaultCISRecordTTL = int64(120)
```
Changed it to const.

SzucsAti commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -1,43 +1,40 @@
 )
 
 type Provider struct {
+	// Mapping zones to cis clients
```
Fixed.

SzucsAti commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -103,7 +101,11 @@ func (p *Provider) Delete(record *iov1.DNSRecord, zone configv1.DNSZone) error {
 	opt := dnsService.NewListAllDnsRecordsOptions()
 	opt.SetType(string(record.Spec.RecordType))
-	opt.SetName(record.Spec.DNSName)
+	// DNS records may have an ending "." character in the DNS name.  For
+	// example, the ingress operator's ingress controller adds a trailing
+	// "." when it creates a wildcard DNS record.
+	dnsName := strings.TrimSuffix(record.Spec.DNSName, ".")
+	opt.SetName(dnsName)
```
If the record was created with a trailing dot, on the CIS side it still appears without the dot. This means ListRecords won't return the records as it will look for records with a trailing dot. It was fixed for update/create here: https://github.com/openshift/cluster-ingress-operator/pull/652, but I believe it should have been added to delete as well.

Miciah replied:
Does that mean that your change here fixes an unreported defect where the operator silently failed to delete old DNS records (since it would look for a record with a trailing ".", fail to find it, and assume there was nothing to delete)?  If so, nice!  And we might want to call that out in a release note.

SzucsAti commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -114,7 +116,7 @@ func (p *Provider) Delete(record *iov1.DNSRecord, zone configv1.DNSZone) error {
 		}
 		if result == nil || result.Result == nil {
-			return fmt.Errorf("delete: invalid result")
+			return fmt.Errorf("delete: ListAllDnsRecords returned nil as result")
```
Changed it.

SzucsAti commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -123,7 +125,7 @@ func (p *Provider) Delete(record *iov1.DNSRecord, zone configv1.DNSZone) error {
 			_, delResponse, err := dnsService.DeleteDnsRecord(delOpt)
 			if err != nil {
-				if delResponse != nil && delResponse.StatusCode != http.StatusNotFound {
+				if delResponse != nil && delResponse.StatusCode != http.StatusNotFound || delResponse == nil {
```
Fixed.

gcs278 commented:
/lgtm but Miciah should review the responses to his comments and approve

Miciah commented:
It's looking really good!  Thanks!
/approve
/lgtm

We have an issue in CI, which is being addressed here: https://github.com/openshift/cluster-ingress-operator/pull/818
I'm putting a hold on this PR while that issue is sorted out so the PR doesn't needlessly rerun CI.  Once the CI issue is resolved, the hold can be removed with `/hold cancel`.  
/hold

jeffnowicki commented:
@Miciah / @gcs278 looks like all tests for https://github.com/openshift/cluster-ingress-operator/pull/818 passed... hopefully it can get reviewed/merged soon so this PR can progress/merge. Thank you.

Miciah commented:
https://github.com/openshift/cluster-ingress-operator/pull/818 merged.  E2E should pass now.  
/hold cancel
/retest-required

Miciah commented:
e2e-aws-operator failed on must-gather.
/test e2e-aws-operator

jeffnowicki commented:
/retest-required

SzucsAti commented:
/retest-required

mkumatag commented:
we need these required flags: `px-approved, qe-approved, docs-approved` @jeffnowicki @SzucsAti

jeffnowicki commented:
@mkumatag I have slack DM'd with @Miciah to get the right people notified so we can get these labels added and PR merged.

ahardin-rh commented:
/label docs-approved

jeffnowicki commented:
@sferich888 @ShudiLi would you please review/add `px-approved` and `qe-approved` labels accordingly?

ShudiLi commented:
/label qe-approved
@jeffnowicki thanks

CFields651 commented:
/label px-approved
